dotnet-authserver/src/TeacherIdentity.AuthServer/Journeys/LegacyTrnJourney.cs

```diff
@@ -1,43 +1,82 @@
 using Microsoft.AspNetCore.Mvc;
-using TeacherIdentity.AuthServer.Models;
+using Microsoft.EntityFrameworkCore;
+using TeacherIdentity.AuthServer.Oidc;
 
 namespace TeacherIdentity.AuthServer.Journeys;
 
 public class LegacyTrnJourney : SignInJourney
 {
-    private readonly CreateUserHelper _createUserHelper;
     private readonly TrnLookupHelper _trnLookupHelper;
 
     public LegacyTrnJourney(
         CreateUserHelper createUserHelper,
         TrnLookupHelper trnLookupHelper,
         HttpContext httpContext,
         IdentityLinkGenerator linkGenerator)
-        : base(httpContext, linkGenerator)
+        : base(httpContext, linkGenerator, createUserHelper)
     {
-        _createUserHelper = createUserHelper;
         _trnLookupHelper = trnLookupHelper;
     }
 
     public bool FoundATrn => AuthenticationState.Trn is not null;
 
-    public Task<IActionResult> CreateOrMatchUserWithTrn(string currentStep) =>
-        _createUserHelper.CreateOrMatchUserWithTrn(this, currentStep);
+    public override async Task<IActionResult> TryCreateUser(string currentStep)
+    {
+        try
+        {
+            var user = await CreateUserHelper.CreateUserWithTrn(AuthenticationState);
+
+            AuthenticationState.OnTrnLookupCompletedAndUserRegistered(user);
+            await AuthenticationState.SignIn(HttpContext);
+
+            if ((!AuthenticationState.TryGetOAuthState(out var oAuthState) || !oAuthState.HasScope(CustomScopes.Trn)) &&
+                AuthenticationState.TrnLookupStatus == TrnLookupStatus.Pending)
+            {
+                await CreateUserHelper.CreateTrnResolutionZendeskTicket(AuthenticationState);
+            }
+        }
+        catch (DbUpdateException dex) when (dex.IsUniqueIndexViolation("ix_users_trn"))
+        {
+            // TRN is already linked to an existing account
+            return await CreateUserHelper.GeneratePinForExistingUserAccount(this, currentStep);
+        }
+
+        return new RedirectResult(GetNextStepUrl(currentStep));
+    }
+
+    public override async Task<RedirectResult> Advance(string currentStep)
+    {
+        if (ShouldPerformTrnLookup(currentStep))
+        {
+            var lookupResult = await _trnLookupHelper.LookupTrn(AuthenticationState);
+            if (lookupResult is not null)
+            {
+                return new RedirectResult(GetStepUrl(Steps.CheckAnswers));
```

Comment:
Haven't updated legacy check answers logic yet so this explicit redirect is still required. Not exactly sure what Legacy changes are required/in-scope. Can update if you like

Reply:
Legacy journey can stay as-is.